zeus/monitor.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (C) 2022 Jae-Won Chung <jwnchung@umich.edu>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for using the Zeus monitor inside training scripts for energy profiling."""
+
+from __future__ import annotations
+
+import time
+import atexit
+import signal
+import tempfile
+import subprocess
+from typing import Generator
+from contextlib import contextmanager
+
+from zeus.analyze import energy as compute_energy
+
+
+class ZeusMonitorContext:
+    """Monitors the energy and time consumption inside a training loop.
+
+    Skip the first `skip_steps` steps and profile for the next `profile_steps`. Does
+    nothing after that, before you call [`reset`][zeus.monitor.ZeusMonitorContext.reset].
+
+    You can check whether profiling is done (i.e., `skip_steps + profile_steps` passed)
+    through [`is_done`][zeus.monitor.ZeusMonitorContext.is_done] and query results through
+    [`energy`][zeus.monitor.ZeusMonitorContext.energy] and
+    [`time`][zeus.monitor.ZeusMonitorContext.time].
+
+    ## Integration example
+
+    ```python
+    zeus_ctx = zeus.monitor.ZeusMonitorContext(skip_steps=10, profile_steps=10)
+
+    for step, (x, y) in enumerate(train_dataloader):
+        print(f"Training step {step}.")
+
+        # Wrap the code range of one training step.
+        # zeus_ctx will ignore the first 10 steps and measure the next 10 steps.
+        with zeus_ctx.step():
+            training_step(x, y)
+
+        # An alternative way if you don't want to use context managers.
+        zeus_ctx.start_step()
+        training_step(x, y)
+        zeus_ctx.finish_step()
+
+        # Check if 20 steps passed and query results.
+        if zeus_ctx.is_done:
+            print(
+                f"{zeus_ctx.profile_steps} training steps "
+                f"consumed {zeus_ctx.energy} Joules in {zeus_ctx.time} seconds."
+            )
+            zeus_ctx.reset()
+    ```
+    """
+
+    def __init__(
+        self,
+        skip_steps: int = 10,
+        profile_steps: int = 10,
+        device_id: int = 0,
+        zeus_monitor_path: str = "zeus_monitor",
+        zeus_monitor_sleep_ms: int = 0,
+        zeus_monitor_log_dir: str | None = None,
+    ) -> None:
+        """Create a Zeus monitor context.
+
+        Args:
+            skip_steps: The number of steps to skip when profiling energy consumption.
+            profile_steps: The number of steps to profile and average over for energy consumption.
```

Comment:
"Averaging" is mentioned here. Since the PR is for profiling per-iteration time, would it be better to add a `property` to compute the _average_ time/energy?